optking/dimerfrag.py

```diff
@@ -739,70 +774,94 @@ def test_orient(NA, NB, printInfo=False) :
-        l.sort()
-        if l in Aatoms:
-            continue
-        n += 1
-        Aatoms.append(l)
-        Aweights.append( ref_length*[1.0] )
-    printInfo: logger.debug(str(Aatoms))
-    printInfo: logger.debug(str(Aweights))
-
-    atom_list = list(range(NB))
-    Batoms = []
-    Bweights = []
-    n = 0
-    while n < NBref:
-        ref_length = min(NBref, choice([1,2,3]))
-        l = sample(atom_list, ref_length)
-        l.sort()
-        if l in Batoms:
-            continue
-        n += 1
-        Batoms.append(l)
-        Bweights.append( ref_length*[1.0])
-    if printInfo: logger.debug(str(Batoms))
-    if printInfo: logger.debug(str(Bweights))
-
-    Itest = DimerFrag(0, Aatoms, 1, Batoms, Aweights, Bweights, Albl, Blbl)
 
+    # Choose a random geometry
     Axyz = np.zeros( (NA,3) )
-    for i in range(NA): # insert random geometries
-        Axyz[i][:] = 5.0+3.0*random(), 3.0*random(), 3.0*random()
     Bxyz = np.zeros( (NB,3) )
+    for i in range(NA):
+        Axyz[i][:] = 6.0+3.0*random(), 3.0*random(), 3.0*random()
     for i in range(NB):
         Bxyz[i][:] = 3.0*random(), 3.0*random(), 3.0*random()
+
+    # Choose # of ref. points not to exceed number of atoms in each fragment.
+    NAref = min(NA,3)
+    NBref = min(NB,3)
+
+    while True:
+        try: # make up some reference points
+            atom_list = list(range(NA))
+            Aatoms = []
+            Aweights = []
+            n = 0
+            while n < NAref:
+                ref_length = min(NAref, choice([1,2,3])) # of atoms used to define ref. pt.
+                l = sample(atom_list, ref_length)  # select random atoms in A
+                l.sort()
+                if l in Aatoms:
+                    continue
+                n += 1
+                Aatoms.append(l)
+                Aweights.append( [uniform(0.1,0.9) for i in range(ref_length)] )
+            if printInfo: logger.debug('FragA atoms:  ' + str(Aatoms))
+            if printInfo: logger.debug('FragA weights ' + str(Aweights))
+    
+            atom_list = list(range(NB))
+            Batoms = []
+            Bweights = []
+            n = 0
+            while n < NBref:
+                ref_length = min(NBref, choice([1,2,3]))
+                l = sample(atom_list, ref_length)
+                l.sort()
+                if l in Batoms:
+                    continue
+                n += 1
+                Batoms.append(l)
+                Bweights.append( [uniform(0.1,0.9) for i in range(ref_length)] )
+            if printInfo: logger.debug('FragB atoms:  ' + str(Batoms))
+            if printInfo: logger.debug('FragB weights:' + str(Bweights))
+    
+            Albl = "A-%d-atoms" % NA
+            Blbl = "B-%d-atoms" % NB
+            Itest = DimerFrag(0, Aatoms, 1, Batoms, Aweights, Bweights, Albl, Blbl)
+            Itest.validate_intcos(Axyz, Bxyz)
+            break
+        except AlgError as error:
+            logger.debug('Trying new reference points')
```

Comment:
I believe you mentioned previously that the collinearity issues only occur if the weights are the same (could have totally misinterpreted). Is there a way to adjust the weights in the case of an AlgError is here? 

Personally I'm a big fan of random input testing, but there should be a few cases where the test generates reference points we know do and don't work.

Reply:
Yes, if you randomly assign weights, then you are for less likely to get coordinates right at discontinuities in the internal definitions, such as result from collinear points.  I learned in an AI book I am reading that you can seed the random module with a particular number and then get reproducible results.  If needed because we getting hassled with weird cases, we'll do that later.